java/test-index/A.java

```diff
@@ -0,0 +1,71 @@
+import ddlogapi.*;
+
+public class A {
+    private final DDlogAPI api;
+
+    A() throws DDlogException {
+        this.api= new DDlogAPI("mylib", 1, false);
+    }
+
+    private DDlogRecCommand[] createCommand(DDlogCommand.Kind kind, String table, int argument0, int argument1) throws DDlogException {
+        DDlogRecord r0 = new DDlogRecord(argument0);
+        DDlogRecord r1 = new DDlogRecord(argument1);
+        DDlogRecord[] rs = new DDlogRecord[2];
+        rs[0] = r0;
+        rs[1] = r1;
+        DDlogRecord o = DDlogRecord.makeStruct(table, rs);
+        int id = this.api.getTableId(table);
+        DDlogRecCommand command = new DDlogRecCommand(kind, id, o);
+        DDlogRecCommand[] result = new DDlogRecCommand[1];
+        result[0] = command;
+        return result;
+    }
+
+    public void run() throws DDlogException {
+        // Insert something in the input table
+        this.api.transactionStart();
+        DDlogRecCommand[] c1 = this.createCommand(DDlogCommand.Kind.Insert, "AI", 0, 1);
+        this.api.applyUpdates(c1);
+        // Apply the changes and get the result
+        DDlogAPI.DDlogCommandVector result = this.api.transactionBatchCommit();
+        assert result.size() == 1;
+        DDlogRecCommand c = result.get(0);
+        assert c.kind() == DDlogCommand.Kind.Insert;
+        DDlogRecord rec = c.value();
+        System.out.println(rec.toString());
+        // Deallocate the result
+        result.dispose();
+
+        // Dump the entire index.
+        this.api.dumpIndex("AII", i -> System.out.println("dump: " + i));
+        rec = new DDlogRecord(0);
+        // Reuse the same key to query the AIII index.
+        this.api.queryIndex("AIII", rec, i -> System.out.println("query 0: " + i));
+        // The key is no longer useful, we need to deallocate it explicitly.
+        rec.dispose();
```

Comment:
Not actually sure why this is necessary. Where are you releasing this memory from? Isn't `rec` just an object to the JVM and will get GC'd accordingly?

Question applies to other `dispose` calls as well.

Reply:
The queryIndex api does not consume the keys, they have to be disposed by the caller.
See also #1092.
The DDlogRecord objects are allocated by the native world and are not GC-ed. That's why they have stringent lifetime requirements. Most often you don't see them because each object is used exactly once (after which the native API usually takes ownership and is responsible for deleting it). This API is different. The Java GC may collect the record, but the record contains a pointer which will never be freed by GC. In a language by C# this is done by objects implementing an IDisposable interface, but Java 8 does not have an equivalent.